ports/stm32/omv_csi.c

```diff
@@ -324,73 +312,74 @@ int omv_csi_config(omv_csi_config_t config) {
 }
 
 // Stop the DCMI from generating more DMA requests, and disable the DMA.
-int omv_csi_abort(omv_csi_t *csi, bool fifo_flush, bool in_irq) {
-    if (DCMI_IS_ACTIVE()) {
-        #if USE_DCMI
-        DCMI->CR &= ~DCMI_CR_ENABLE;
-        #endif
-        #if USE_DMA
-        if (in_irq) {
-            HAL_DMA_Abort_IT(&csi->dma);
-        } else {
-            HAL_DMA_Abort(&csi->dma);
-        }
-        HAL_NVIC_DisableIRQ(DMA2_Stream1_IRQn);
-        #endif
-        #if USE_MDMA
-        if (!in_irq) {
-            HAL_MDMA_Abort(&csi->mdma0);
-            HAL_MDMA_Abort(&csi->mdma1);
-        }
-        HAL_MDMA_DeInit(&csi->mdma0);
-        HAL_MDMA_DeInit(&csi->mdma1);
-        #endif
-        #if USE_DCMI
-        __HAL_DCMI_DISABLE_IT(&csi->dcmi, DCMI_IT_FRAME);
-        __HAL_DCMI_CLEAR_FLAG(&csi->dcmi, DCMI_FLAG_FRAMERI);
-        #else
-        if (!csi->mipi_if) {
-            HAL_DCMIPP_PIPE_Stop(&csi->dcmi, DCMIPP_PIPE);
-        } else {
-            HAL_DCMIPP_CSI_PIPE_Stop(&csi->dcmi, DCMIPP_PIPE, DCMIPP_VIRTUAL_CHANNEL0);
-        }
-        for (size_t i=0; i<DCMIPP_NUM_OF_PIPES; i++) {
-            csi->dcmi.PipeState[i] = HAL_DCMIPP_PIPE_STATE_RESET;
-        }
-        #endif
-        csi->first_line = false;
-        csi->drop_frame = false;
-        csi->last_frame_ms = 0;
-        csi->last_frame_ms_valid = false;
+static int stm_csi_abort(omv_csi_t *csi, bool fifo_flush, bool in_irq) {
+    if (!DCMI_IS_ACTIVE()) {
+        return 0;
     }
 
-    if (csi->fb) {
-        if (fifo_flush) {
-            framebuffer_flush_buffers(csi->fb, true);
-        } else if (!csi->disable_full_flush) {
-            framebuffer_flush_buffers(csi->fb, false);
-        }
+    #if USE_DCMI
+    DCMI->CR &= ~DCMI_CR_ENABLE;
+    #endif
+
+    #if USE_DMA
+    if (in_irq) {
+        HAL_DMA_Abort_IT(&csi->dma);
+    } else {
+        HAL_DMA_Abort(&csi->dma);
+    }
+    HAL_NVIC_DisableIRQ(DMA2_Stream1_IRQn);
+    #endif
+
+    #if USE_MDMA
+    if (!in_irq) {
+        HAL_MDMA_Abort(&csi->mdma0);
+        HAL_MDMA_Abort(&csi->mdma1);
     }
+    HAL_MDMA_DeInit(&csi->mdma0);
+    HAL_MDMA_DeInit(&csi->mdma1);
+    #endif
+
+    #if USE_DCMI
+    __HAL_DCMI_DISABLE_IT(&csi->dcmi, DCMI_IT_FRAME);
+    __HAL_DCMI_CLEAR_FLAG(&csi->dcmi, DCMI_FLAG_FRAMERI);
+    #else
+    if (!csi->mipi_if) {
+        HAL_DCMIPP_PIPE_Stop(&csi->dcmi, DCMIPP_PIPE);
+    } else {
+        HAL_DCMIPP_CSI_PIPE_Stop(&csi->dcmi, DCMIPP_PIPE, DCMIPP_VIRTUAL_CHANNEL0);
+    }
+    for (size_t i=0; i<DCMIPP_NUM_OF_PIPES; i++) {
+        csi->dcmi.PipeState[i] = HAL_DCMIPP_PIPE_STATE_RESET;
+    }
+    #endif
 
     return 0;
 }
 
 uint32_t omv_csi_get_clk_frequency() {
```

Comment:
Should get/set clk freq have a csi object? I noticed they seem to be excluded. I guess they are always the primary? It may be better to have overrides like everything else. Like, for example, with the Lepton+PAG7936 board the lepton has it's own 25MHz clock source. That freq could be hardcoded into the firmware to be returned when it's an aux sensor. That said, I don't think anyone will actually care about this ever... so, the current approach is probably fine until there's an issue requiring it to be changed.

Reply:
I've thought about that, but quickly realized it will only lead to set_clk getting called more times than it is right now. Also, the only reason to override those is if a sensor requires a different clock source, but you won't know that until the sensor is clocked and detected.